cli/CommandLineParser.php

```diff
@@ -0,0 +1,277 @@
+<?php
+declare(strict_types=1);
+
+class CommandLineParser {
+	/** @var array<string,Option> */
+	private array $options;
+	/** @var array<string,array{defaultInput:?string[],required:?bool,aliasUsed:?string,values:?string[]}> */
+	private array $inputs;
+
+	/** Adds an option that produces an error message if not set. */
+	public function addRequiredOption(string $name, Option $option): static {
+		$this->inputs[$name] = [
+			'defaultInput' => null,
+			'required' => true,
+			'aliasUsed' => null,
+			'values' => null,
+		];
+		$this->options[$name] = $option;
+
+		return $this;
+	}
+
+	/**
+	 * Adds an optional option.
+	 * @param string $defaultInput If not null this value is received as input in all cases where no
+	 *  user input is present. e.g. set this if you want an option to always return a value.
+	 */
+	public function addOption(string $name, Option $option, string $defaultInput = null) :static {
+		$this->inputs[$name] = [
+			'defaultInput' => is_string($defaultInput) ? [$defaultInput] : $defaultInput,
+			'required' => null,
+			'aliasUsed' => null,
+			'values' => null,
+		];
+		$this->options[$name] = $option;
+
+		return $this;
+	}
+
+	/**
+	 * @template T
+	 * @param class-string<T> $target
+	 * @return T
+	 */
+	public function parse(string $target) {
+		global $argv;
+
+		$output = new $target();
+		$output->errors = [];
```

Comment:
If you could find a way to avoid dynamic properties, that would be nice. I am trying to get rid of them throughout our codebase, as they are difficult to maintain over time (how to know which ones exist, for instance) compared to more strongly-typed approaches.

Maybe by having the some predefined children of the class `Option`?

Reply:
With [09552c7](https://github.com/FreshRSS/FreshRSS/pull/6099/commits/09552c7e164f1f3599d09a29f0b21ece4c918eff) properties are now set in a 'Definition' `Class` for all commands, this 'Definition' `Class` is now given to the `parse()` method on `CommandLineParser` letting us move away from using dynamic properties.

I had intended to go ahead with something along these lines anyway. I ended up using the `$parser->parse(stdClass::class)` format as it felt like it worked a little more cleanly with how things stand. Although, I certainly agree that moving away from dynamic properties is desirable.

Reply:
Improved a bit more in https://github.com/KasimirCash/FreshRSS/pull/1
Example: `$options = new ActualizeUserDefinition();` instead of `$options = $parser->parse(ActualizeUserDefinition::class);`